src/Microsoft.FeatureManagement/FeatureManagementBuilderExtensions.cs

```diff
@@ -39,6 +40,41 @@ public static class FeatureManagementBuilderExtensions
             return builder;
         }
 
+        /// <summary>
+        /// Adds a <see cref="VariantServiceProvider{TService}"/> to the feature management system.
+        /// </summary>
+        /// <param name="builder">The <see cref="IFeatureManagementBuilder"/> used to customize feature management functionality.</param>
+        /// <param name="featureName">The variant feature flag used to assign variants. The <see cref="VariantServiceProvider{TService}"/> will return different implementations of TService according to the assigned variant.</param>
+        /// <returns>A <see cref="IFeatureManagementBuilder"/> that can be used to customize feature management functionality.</returns>
+        public static IFeatureManagementBuilder AddVariantServiceProvider<TService>(this IFeatureManagementBuilder builder, string featureName) where TService : class
+        {
+            if (string.IsNullOrEmpty(featureName))
+            {
+                throw new ArgumentNullException(nameof(featureName));
+            }
+
+            if (builder.Services.Any(descriptor => descriptor.ServiceType == typeof(IFeatureManager) && descriptor.Lifetime == ServiceLifetime.Scoped))
+            {
+                builder.Services.AddScoped<IVariantServiceProvider<TService>>(sp => new VariantServiceProvider<TService>(
```

Comment:
Does this service make sense as a scoped service? From my understanding- the variant service will be a Singleton, right? Or, if the variant service could be scoped, then perhaps this should check the descriptor.Lifetime of one of the variant services instead of FeatureManager.

Reply:
VariantServiceProvider will consume feature manager. If the feature manager is scoped and we register the variant service provider as singleton, an exception will be thrown. 

> check the descriptor.Lifetime of one of the variant services instead of FeatureManager.
This will fail if we cannot find variant services in the service collection, because users may register their implementation after calling our API like:
```C#
services.AddFeatureManagement()
            .AddVariantServiceProvider<IAlgorithm>("MyFeature");  
services.AddSingleton<IAlgorithm, AlgorithmAlpha>();
```

Reply:
To clarify, I'm curious about the actual variant services- not the provider. Ie:
```
services.AddFeatureManagement()
        .AddVariantServiceProvider<IAlgorithm>("MyFeature");  
services.AddSingleton<IAlgorithm, AlgorithmAlpha>(); // This 
```
vs
```
services.AddFeatureManagement()
        .AddVariantServiceProvider<IAlgorithm>("MyFeature");  
services.AddScoped<IAlgorithm, AlgorithmAlpha>(); // vs this
```

Reply:
If the user registers variant service as scoped and registers feature management as singleton, then the user mess up the service lifetime. Scoped variant service will be consumed by the singleton variant service provider, an exception will occur.
```C#
services.AddFeatureManagement()
        .AddVariantServiceProvider<IAlgorithm>("MyFeature");  
services.AddScoped<IAlgorithm, AlgorithmAlpha>();
```

The variant services are registered by user, so they should choose the right lifetime and the correct method to register feature management (AddScopedFeatureManagement or AddFeatureManagement).

For me, this is the similar case for targeting context accessor, before we provide `WithTargeting` API, user need to register `ITargetingContextAccessor` by themselves and they can register it as scoped/transient whatever they want. We didn't register the targeting filter or feature manager as scoped/transient if we find `ITargetingContextAccessor` is registered as scoped. Instead, if user really wants to use a scoped `ITargetingContextAccessor`, they should call `AddScopedFeatureManagement` to register the whole feature management as scoped.